bumble/hid.py

```diff
@@ -242,37 +297,224 @@ def on_l2cap_channel_open(self, l2cap_channel: l2cap.ClassicChannel) -> None:
+                else:
+                    self.send_handshake_message(Message.Handshake.ERR_INVALID_PARAMETER)
+
+            elif ret.status == self.GetSetReturn.REPORT_ID_NOT_FOUND:
+                self.send_handshake_message(Message.Handshake.ERR_INVALID_REPORT_ID)
+            elif ret.status == self.GetSetReturn.ERR_INVALID_PARAMETER:
+                self.send_handshake_message(Message.Handshake.ERR_INVALID_PARAMETER)
+            elif ret.status == self.GetSetReturn.ERR_UNSUPPORTED_REQUEST:
+                self.send_handshake_message(Message.Handshake.ERR_UNSUPPORTED_REQUEST)
+        else:
+            logger.debug("GetReport callback not registered !!")
+            self.send_handshake_message(Message.Handshake.ERR_UNSUPPORTED_REQUEST)
+
+    def register_get_report_cb(self, cb):
+        self.get_report_cb = cb
+        logger.debug("GetReport callback registered successfully")
+
+    def handle_set_report(self, pdu: bytes):
+        if self.set_report_cb != None:
+            report_type = pdu[0] & 0x03
+            report_id = pdu[1]
+            report_data = pdu[2:]
+            report_size = len(pdu[1:])
+            ret = self.set_report_cb(
+                report_id, report_type, report_size, report_data
+            )  #  type: ignore
+            if ret.status == self.GetSetReturn.SUCCESS:
+                self.send_handshake_message(Message.Handshake.SUCCESSFUL)
+            elif ret.status == self.GetSetReturn.ERR_INVALID_PARAMETER:
+                self.send_handshake_message(Message.Handshake.ERR_INVALID_PARAMETER)
+            elif ret.status == self.GetSetReturn.REPORT_ID_NOT_FOUND:
+                self.send_handshake_message(Message.Handshake.ERR_INVALID_REPORT_ID)
+            else:
+                self.send_handshake_message(Message.Handshake.ERR_UNSUPPORTED_REQUEST)
+        else:
+            logger.debug("SetReport callback not registered !!")
+            self.send_handshake_message(Message.Handshake.ERR_UNSUPPORTED_REQUEST)
+
+    def register_set_report_cb(self, cb):
+        self.set_report_cb = cb
+        logger.debug("SetReport callback registered successfully")
+
+    def handle_get_protocol(self, pdu: bytes):
+        ret = self.GetSetStatus()
+        if self.get_protocol_cb != None:
+            ret = self.get_protocol_cb()  #  type: ignore
+            if ret.status == self.GetSetReturn.SUCCESS:
+                self.send_control_data(Message.ReportType.OTHER_REPORT, ret.data)
+                return
+        else:
+            logger.debug("GetProtocol callback not registered !!")
+
+        self.send_handshake_message(Message.Handshake.ERR_UNSUPPORTED_REQUEST)
+
+    def register_get_protocol_cb(self, cb):
+        self.get_protocol_cb = cb
+        logger.debug("GetProtocol callback registered successfully")
+
+    def handle_set_protocol(self, pdu: bytes):
+        ret = self.GetSetStatus()
+        if self.set_protocol_cb != None:
```

Comment:
See earlier comment about testing for None at the start and returning.

Reply:
Resolved